cpp/src/arrow/scalar.cc

```diff
@@ -1149,7 +1149,7 @@ Result<std::shared_ptr<Scalar>> Scalar::CastTo(std::shared_ptr<DataType> to) con
   if (is_valid) {
     out->is_valid = true;
     ToTypeVisitor unpack_to_type{*this, to, out.get()};
-    RETURN_NOT_OK(VisitTypeInline(*to, &unpack_to_type));
+    RETURN_NOT_OK(VisitScalarTypeInline(*to, &unpack_to_type));
```

Comment:
Hmm, we should definitely define a `RunEndEncodedScalar` instead of hacking our way around its lack, IMHO.

Reply:
That's what I've been working on for a few days now. I'm also adding JSON converters in the process and rewriting the array builder.

Reply:
And I totally agree. Not having the Scalar is not something we can really get away with.